### PR TITLE
Add ESP32S3

### PR DESCRIPTION
**Pull request by JurajSadel**
This PR is built on top of https://github.com/esp-rs/esp-flasher-stub/pull/3 so technically adds support for ESP32 and ESP32S3.
Contains also an initial cleanup and refactor.
Updates esp-hal versions.

### CODE REVIEW

File: src/commands.rs [hunk 1]
```diff
@@ -162,52 +171,34 @@ pub struct ReadFlashParams {
...
 
-// Size of sesponse without data reference
-pub const RESPONSE_SIZE: usize = 10;
-
 impl<'a> Response<'a> {
-    pub fn new(cmd: Code) -> Self {
+    pub fn new(cmd: CommandCode) -> Self {
         Response {
-            direction: Direction::Out,
+            direction: 1,
```
Comment (igrr) [inline src/commands.rs:196]: I wonder if the version with enums is not more readable then the version with an integer constant?

File: src/serial_io.rs [hunk 2]
```diff
@@ -1,41 +1,110 @@
-use esp32c3_hal::{
-    interrupt,
-    interrupt::CpuInterrupt,
-    pac::{self, UART0},
-    prelude::*,
-    serial::Instance,
-    Cpu, Serial,
+use heapless::Deque;
+#[cfg(any(target_arch = "riscv32"))]
```
Comment (bjoernQ) [inline src/serial_io.rs:2]: we could use critical-section now - no need to worry about the target architecture then

File: src/dprint.rs [hunk 3]
```diff
@@ -4,21 +4,19 @@
 //! This is unsafe! It is asynchronous with normal UART1 usage and
 //! interrupts are not disabled.
 
-use esp32c3_hal::{pac::UART1, prelude::nb};
```
Comment (MabezDev) [inline src/dprint.rs:7]: We can remove this module completely now, and use esp_println with the log crate instead. (If we use the log crate we can optimize logs out later by setting a feature)

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+// Re-export the correct HAL based on which feature is active
+#[cfg(feature = "esp32")]
+pub use esp32_hal as hal;
+#[cfg(feature = "esp32c3")]
+pub use esp32c3_hal as hal;
+#[cfg(feature = "esp32s3")]
+pub use esp32s3_hal as hal;
+// Due to a bug in esp-hal this MUST be included in the root.
+#[cfg(feature = "esp32c3")]
+pub use hal::interrupt;
```
Comment (MabezDev) [inline src/lib.rs:12]: Which bug is this? Is there an issue?

File: src/serial_io.rs [hunk 5]
```diff
@@ -1,41 +1,110 @@
...
-        unsafe { RX_QUEUE.enqueue(data).unwrap() };
+        unsafe { RX_QUEUE.push_back(data).unwrap() };
     }
 
     uart.int_clr.write(|w| w.rxfifo_full_int_clr().set_bit());
-    interrupt::clear(Cpu::ProCpu, CpuInterrupt::Interrupt3);
+}
+
+#[cfg(feature = "esp32")]
+pub fn enable_uart0_rx_interrupt() {
```
Comment (MabezDev) [inline src/serial_io.rs:48]: We have the ability to enable serial interrupts in esp-hal now, so we can remove these functions now. Though we may still need the interrupt mapping parts.

File: src/protocol.rs [hunk 6]
```diff
@@ -508,397 +536,3 @@ mod slip {
         io.send(&[0xC0]);
     }
 }
-
```
[the remark this answers is not in the document]
Comment (JurajSadel) [reply src/protocol.rs:511]: Let's address them in another PR.

File: src/commands.rs (hunk 1 above)
Comment (JurajSadel) [reply src/commands.rs]: You are right, thank you!

File: src/dprint.rs (hunk 3 above)
Comment (JurajSadel) [reply src/dprint.rs:7]: Let's address this it another PR.

File: src/serial_io.rs (hunk 5 above)
Comment (JurajSadel) [reply src/serial_io.rs]: It's referring to https://github.com/esp-rs/esp-hal/issues/197

File: src/lib.rs (hunk 4 above)
Comment (JurajSadel) [reply src/lib.rs:12]: It's referring to https://github.com/esp-rs/esp-hal/issues/197

**Comment by JurajSadel** [comment]
@bjoernQ @igrr @MabezDev thank you for doing review, I pushed second round of cleanup.

FYI @DNedic, @dobairoland, @radimkarnis

**Review by bjoernQ — APPROVED** [review]
LGTM - some improvements will be addressed in subsequent PRs

**Pull request merged**